### PR TITLE
feat(tests): mock tests part 1 - create generalized and random mock dataclasses

### PR DESCRIPTION
**Pull request by Aurelien9Code**
UPDATE : This PR implements a Mock class in the tests for Associations, Events, Locations, Users, ReferenceLists for the backend tests.

My implementation could contain circular dependency, explained in #132 Annoucement. 

I aslo removed any kind of randomization, as it could lead to different test outputs. 

All the edge cases are impemented, please feel free to add some in your new dataclasses !

### CODE REVIEW

**Review by Zafouche — COMMENTED** [review]
Great initiative @Aurelien9Code! Randomness can be both good to make sure every edge case is covered, however it could mean the CI passes one time while the other time, it doesn't, so that could cause problems.

**Review by armouldr — CHANGES_REQUESTED** [review]
Overall good addition to the project that's gonna facilitate testing and especially make the test clearer. However I think that you should make it easier for someone to get a non-random object from these "MockObject" classes, because for now creating a non-random object with your classes is the same as creating the object directly.

You should also change change the boolean value to true at line 85 of MockAssociation.kt as I believe this uses more memory than necessary.

To summarise, there's just some little arrangements to make and I'll approve your PR :)

File: app/src/test/java/com/android/unio/model/association/MockAssociation.kt [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        url: String = getRandomUrl(),
+        name: String = getRandomName(),
+        fullName: String = getRandomFullName(),
+        category: AssociationCategory = getRandomAssociationCategory(),
+        description: String = getRandomDescription(),
+        members: List<User> =
+            if (userDependency) {
+              emptyList()
+            } else {
+              MockUser.createAllMockUsers(associationDependency = false)
```
[the remark this answers is not in the document]
Comment (Aurelien9Code) [reply app/src/test/java/com/android/unio/model/association/MockAssociation.kt:85]: Yes you are right, it's just a minor error and I will change it to "true"

File: app/src/test/java/com/android/unio/model/association/MockAssociation.kt [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+/**
+ * MockAssociation class provides sample instances of the Association data class for testing
+ * purposes.
+ */
+class MockAssociation {
+  companion object {
+
+    /** Helper function to generate a random UID for an association. */
+    private fun getRandomUid(): String = "assoc${Random.nextInt(1000, 9999)}"
```
[the remark this answers is not in the document]
Comment (Aurelien9Code) [reply app/src/test/java/com/android/unio/model/association/MockAssociation.kt:17]: I didn't catch your sentence

File: app/src/test/java/com/android/unio/model/association/MockAssociation.kt [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    /** Helper function to generate a random description for an association. */
+    private fun getRandomDescription(): String {
+      val descriptions =
+          listOf(
+              "A vibrant community focused on learning and growth.",
+              "Committed to bringing people together for various events and activities.",
+              "An association dedicated to promoting sustainability.",
+              "An enthusiastic group of sports lovers and fitness advocates.",
+              "Encouraging cultural awareness and engagement in the arts.",
+              "A space for tech enthusiasts to collaborate and innovate.")
```
[the remark this answers is not in the document]
Comment (Aurelien9Code) [reply app/src/test/java/com/android/unio/model/association/MockAssociation.kt:51]: I think then it would be better to do it inside a test calling my mock, and overwriting the description parameter ; )

File: app/src/test/java/com/android/unio/model/event/MockEvent.kt [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        title: String = "Event ${Random.nextInt(1000)}",
+        organisers: List<Association> = MockAssociation.createAllMockAssociations(),
+        taggedAssociations: List<Association> = MockAssociation.createAllMockAssociations(),
+        image: String = "https://example.com/event_image_${Random.nextInt(1000)}.png",
+        description: String = "This is a detailed description of a random event.",
+        catchyDescription: String = "Catchy event tagline!",
+        price: Double = Random.nextDouble(5.0, 100.0),
+        date: Timestamp = Timestamp(Date(Date().time + Random.nextLong(86400000, 604800000))),
+        location: Location = MockLocation.createMockLocation(),
+        types: List<EventType> = listOf(getRandomEventType())
```
[the remark this answers is not in the document]
Comment (Aurelien9Code) [reply app/src/test/java/com/android/unio/model/event/MockEvent.kt:41]: Good catch !

File: app/src/test/java/com/android/unio/model/firestore/MockReferenceList.kt [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+  override val list: StateFlow<List<T>> = _list
+
+  override fun add(uid: String) {}
+
+  override fun addAll(uids: List<String>) {}
+
+  override fun requestAll(onSuccess: () -> Unit) {
+    onSuccess()
+  }
+}
```
[the remark this answers is not in the document]
Comment (Aurelien9Code) [reply app/src/test/java/com/android/unio/model/firestore/MockReferenceList.kt]: I think it's already done.. I don't see it

**Review by armouldr — DISMISSED** [review]

**Comment by Aurelien9Code** [comment]
Yes i didn't finish haha

**Comment by Aurelien9Code** [comment]
Please check https://github.com/SwEnt-Group13/Unio/pull/123/#issue-2627568739

Sonar failing condition will be fixed in next big refactor of the tests

**Review by armouldr — APPROVED** [review]
Looks all good to me now ! Would be great to have more line coverage but I think it will increase as we incorporate your new mock methods into our tests :)
